staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -196,18 +196,52 @@ func (c *CloudNodeLifecycleController) MonitorNodes(ctx context.Context) {
 	}
 }
 
+// getProviderID returns the provider ID for the node. If Node CR has no provider ID,
+// it will be the one from the cloud provider.
+func (c *CloudNodeLifecycleController) getProviderID(ctx context.Context, node *v1.Node) (string, error) {
+	if node.Spec.ProviderID != "" {
+		return node.Spec.ProviderID, nil
+	}
+
+	if instanceV2, ok := c.cloud.InstancesV2(); ok {
+		metadata, err := instanceV2.InstanceMetadata(ctx, node)
+		if err != nil {
+			if err == cloudprovider.InstanceNotFound {
+				return "", nil
+			}
+			return "", err
+		}
+		return metadata.ProviderID, nil
+	}
+
+	providerID, err := cloudprovider.GetInstanceProviderID(ctx, c.cloud, types.NodeName(node.Name))
+	if err != nil {
+		if err == cloudprovider.InstanceNotFound {
+			return "", nil
```

Comment:
instead of returning nil here, can we return `"", err` and check for `cloudprovider.InstanceNotFound` from  `ensureNodeExistsByProviderID` and `shutdownInCloudProvider`?

Reply:
I have nothing against, it's just the code duplication that was pointed out yesterday and I tried to remove. The next time we have a function that uses `getProviderID` it will need to again have the `if err == cloudprovider.InstanceNotFound`